ex7/s_tokida/ex7.py

```diff
@@ -0,0 +1,252 @@
+    # print('data.shape', data.shape)
+    start = time.time()
+
+    if args.init == 'random':
+        print('random')
+        pi, mu, sigma = init_random(data, num_c)
+
+    elif args.init == 'kmeans':
+        print('kmeans')
+        centroid = kmeans.init_random(data, num_c)
+        cluster, centroid = kmeans.k_means(data, num_c, centroid)
+        cov = np.zeros((num_c, 2, 2))
+        for i in range(num_c):
+            cov[i] = np.cov(data[np.where(cluster == i)].T)
+        pi = np.full(num_c, 1 / num_c)
+        mu = centroid 
+        sigma = cov
+
+    elif args.init == 'kmeans++':
+        print('kmeans++')
+        centroid = kmeans.k_means_plusplus(data, num_c)
+        cluster, centroid = kmeans.k_means(data, num_c, centroid)
+        cov = np.zeros((num_c, 2, 2))
+        for i in range(num_c):
+            cov[i] = np.cov(data[np.where(cluster == i)].T)
+        pi = np.full(num_c, 1 / num_c)
+        mu = centroid 
+        sigma = cov
+    
+    likelihoods, pi, mu, sigma = em_algorithm(data, pi, mu, sigma)
+
+    # time
+    calc_time = time.time() - start
+    print(f"calc_time({args.init}): {calc_time:.5f}")
+
+    colors = ['y', 'm', 'c', 'r', 'g', 'b']
+
+    if data.shape[1] == 1:
+        fig, ax = plt.subplots(figsize = (7,6))
+        ax.set(xlabel = 'x', ylabel = 'probability density', title = f'{filename} K={num_c}')
+
+        pos = np.linspace(np.min(data) - 1, np.max(data) + 1, 100)
+        z = np.zeros(100)
+        for k in range(num_c):
+            z += pi[k] * multivariate_normal.pdf(pos, mu[k], sigma[k])
+
+        ax.scatter(data, np.zeros_like(data), edgecolors='blueviolet', facecolor='None', label='Observed data')
+        plt.plot(pos, z, c='orange', label='GMM')
+        ax.scatter(mu, np.zeros(num_c), marker = '*', s=100, c = 'k', label = 'Centroid' )
+        plt.legend()
+        plt.tight_layout()
+        # plt.savefig('fig/' + f'{filename}_k{num_c}_{args.init}.png')
+        plt.show()
+        plt.close()
+
+        # plot likelihood
+        fig, ax = plt.subplots()
+        ax.set(xlabel = 'Iteration', ylabel = 'Log Likelihood', title = f'{filename} likelihood')
+        
+        plt.plot(np.arange(0, len(likelihoods), 1), likelihoods)
+        # plt.savefig('fig/' + f'{filename}_lh.png')
+        plt.show()
+
+    if data.shape[1] == 2:
+
+        x1 = np.linspace(np.min(data[:, 0]) - 1, np.max(data[:, 0]) + 1, 100)
+        x2 = np.linspace(np.min(data[:, 1]) - 1, np.max(data[:, 1]) + 1, 100)
+        x1, x2 = np.meshgrid(x1, x2)
+        pos = np.dstack((x1, x2))
+
+        prob = np.array([mix_gaussian(x_pos, pi, mu, sigma)[0] for x_pos in pos])
+    
+        fig, ax = plt.subplots(figsize = (7,6))
+        ax.set(xlabel = '$x_1$', ylabel = '$x_2$', title = f'{filename} K={num_c}')
+
+        ax.scatter(data[:, 0], data[:, 1], edgecolors='blueviolet', facecolor='None', label = 'Observed data')
+        ax.scatter(mu[:, 0], mu[:, 1], marker = '*', s=100, c = 'k', label = 'Centroid' )
```

Comment:
例えばこのあたりはスペースの有無を統一していると美しいです（細かいですけどね）
```suggestion
        ax.scatter(data[:, 0], data[:, 1], edgecolors='blueviolet', facecolor='None', label='Observed data')
        ax.scatter(mu[:, 0], mu[:, 1], marker='*', s=100, c='k', label='Centroid')
```
参考：https://qiita.com/simonritchie/items/bb06a7521ae6560738a7#%E3%82%B9%E3%83%9A%E3%83%BC%E3%82%B9%E3%81%AE%E5%85%A5%E3%82%8C%E6%96%B9
これも自動整形してくれるリンターなどを導入するとラクになります．
https://qiita.com/tsu_0514/items/2d52c7bf79cd62d4af4a